### PR TITLE
#166790009 Add Search Functionality And Custom filtering Based on Parameters

### PR DESCRIPTION
**Pull request by elemanhillary-zz**
#### What does this PR do?
- Adds search functionality and custom filtering based on parameters

#### Description of Task to be completed
- Enable user to filter articles by author, tag, title and keywords
- Enable a user to get an article that matches only the criteria that matches the criteria for the article they want.

#### How should this be manually tested?
```bash
git clone https://github.com/andela/ah-92explorers-backend.git
cd ah-92explorers-backend
git checkout ft-search-functionality-166790009
npm install
npm start
```
- Example Usage
*Open Favourite Web Browser* and open
`localhost:3000/api/articles?tag=javascript`

#### Any background context you want to provide?
- None
#### What are the relevant pivotal tracker stories?
[#166790009](https://www.pivotaltracker.com/story/show/166790009)
#### Screenshots (if appropriate)
Imgur
#### Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I don't have more than 3 major commits on this PR

### CODE REVIEW

**Review by niyoceles — CHANGES_REQUESTED** [review]
Squash all commits to a single commit.

**Review by niyoceles — DISMISSED** [review]

**Review by Hervera — DISMISSED** [review]
Ready to be marged

**Review by kodek-sleuth — DISMISSED** [review]

**Review by kodek-sleuth — DISMISSED** [review]

**Review by Hervera — DISMISSED** [review]
Ready to be merged

**Review by IsaiahRn — DISMISSED** [review]

File: controllers/search.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  } = req.query;
+  try {
+    const articlesResults = await articles.findAll();
+    index.addObjects(articlesResults);
+    index.search(`${tag || author || title || keyword}`, (err, results) => {
+      const sanitizedResults = sanitize(results.hits);
+      if (sanitizedResults.length !== 0) {
+        return res.status(200).json({ results: sanitizedResults, });
+      }
+      return res.status(204).json({ error: 'no results' });
```
Comment (elemanhillary-zz) [inline controllers/search.js]: change `204` to `404`

File: tests/search.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    chai.request(app)
+      .get('/api/articles?keyword=how')
+      .end((err, res) => {
+        expect(Object.prototype.toString.call(res.body.results)).to.be.equal('[object Array]');
+        assert.notEqual(res.body.results.length, 0);
+        expect(res.body.results[0].title).to.be.equal('How to train your dragon');
+        done();
+      });
+  });
+  it('should return 204 on empty search results', (done) => {
```
Comment (elemanhillary-zz) [inline tests/search.test.js]: change string `204` to `404`

**Review by niyoceles — DISMISSED** [review]

**Review by IsaiahRn — APPROVED** [review]

**Review by IsaiahRn — CHANGES_REQUESTED** [review]
Kindly. squash your commits

**Comment by elemanhillary-zz** [comment]
> Kindly. squash your commits

had forgotten that hehehehe

**Review by IsaiahRn — APPROVED** [review]
It can be merged

**Review by niyoceles — APPROVED** [review]

**Review by Hervera — CHANGES_REQUESTED** [review]
In files changed, I don't see the code of your controller

**Review by niyoceles — APPROVED** [review]

**Review by niyoceles — APPROVED** [review]

**Review by IsaiahRn — APPROVED** [review]

**Review by Hervera — APPROVED** [review]
Ready to be merged

**Pull request merged**